### PR TITLE
Add Counter check

### PR DESCRIPTION
**Pull request by caryjpaepen**
This is to address https://github.com/cloudflare/pint/issues/655.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,4 @@
 /memprofile.out
 /dist/
 /.vscode/
+.idea
```
Comment (caryjpaepen) [inline .gitignore:6]: Should I be adding this here (it makes sense to add it if it's just our repo)? How likely are we merging this upstream?

File: README.md [hunk 2]
```diff
@@ -14,6 +14,12 @@ Changelog is kept at [docs/changelog.md](/docs/changelog.md).
 
 Check [examples](/docs/examples) dir for sample config files.
 
+## Setup
```
Comment (thejuan) [inline README.md]: I would clarify the environment this is for, this projects team is not on windows

File: internal/checks/promql_counter.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+			isCounterMap.values[s.Name] = isCounter
+		}
+
+		if isCounterMap.values[s.Name] && !parentUsesAllowedFunction {
+			allowedFuncString := "`" + strings.Join(AllowedCounterFuncNames, "`, `") + "`"
+
+			p := exprProblem{
+				expr: node.Expr,
+				text: fmt.Sprintf("Counter metric `%s` should be used with one of following functions: (%s).", s.Name, allowedFuncString),
+				// There can be valid edge cases like a recording rule: `foo{label="value"}` or being constrained to use a counter as an info metric for joining.
```
Comment (thejuan) [inline internal/checks/promql_counter.go:111]: Can we do more to ignore the edge cases?

File: .gitignore (hunk 1 above)
Comment (thejuan) [reply .gitignore:6]: Think this is fine

File: internal/checks/promql_counter.go (hunk 3 above)
Comment (thejuan) [reply internal/checks/promql_counter.go:111]: Chatted with Cary, not without impacting the quality of positives. People can use pragma to disable as needed

**Review by thejuan — APPROVED** [review]

File: docs/checks/promql/counter.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+---
+
+# promql/counter
+
+This check inspects counter metrics used in queries to verify that the raw counter value is not used in calculations.
+
+A counter should be wrapped by one of the following functions:
+
+- `rate`, `irate`, `increase` - [performing rate calculations while handling counter resets](https://promlabs.com/blog/2021/01/29/how-exactly-does-promql-calculate-rates/)
+- `count`, `count_over_time`, `absent`, `absent_over_time` - these are common and valid use cases of any raw metric
```
Comment (djluck) [inline docs/checks/promql/counter.md]: Do we need to think about allowing any other `<agg>_over_time()` functions?

File: docs/checks/promql/counter.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+grand_parent: Documentation
+---
+
+# promql/counter
+
+This check inspects counter metrics used in queries to verify that the raw counter value is not used in calculations.
+
+A counter should be wrapped by one of the following functions:
+
+- `rate`, `irate`, `increase` - [performing rate calculations while handling counter resets](https://promlabs.com/blog/2021/01/29/how-exactly-does-promql-calculate-rates/)
```
Comment (djluck) [inline docs/checks/promql/counter.md:13]: There is specific guidance to disallow `irate` in alerts (see https://www.robustperception.io/avoid-irate-in-alerts/), is it possible to allow `irate` for recording rules only?

File: docs/checks/promql/counter.md (hunk 4 above)
Comment (caryjpaepen) [reply docs/checks/promql/counter.md]: found another one - present_over_time

**Review by djluck — APPROVED** [review]

**Pull request closed without merging**